futures-util/src/future/join_all.rs

```diff
@@ -86,25 +101,30 @@ where
     ) -> Poll<Self::Output> {
         let mut all_done = true;
 
-        for elem in self.as_mut().elems.iter_mut() {
+        let elems: &mut [ElemState<F>] = unsafe {
```

Comment:
seems like we need `get_pin_ref(index: usize)` and `get_pin_mut(index: usize)`

Reply:
(this would address your point (2) above)

Reply:
Yep, I was mostly thinking to throw together a quick prototype here to get this done then open a PR to Rust once it's been validated a little. (For ergonomics here I was planning something like `iter_pin_mut(self: Pin<&mut [T]>) -> impl Iterator<Item = Pin<&mut T>> + '_`).

Reply:
sgtm